src/mdns/mdns_avahi.cpp

```diff
@@ -851,33 +851,39 @@ std::string PublisherAvahi::MakeFullName(const std::string &aName)
 
 void PublisherAvahi::SubscribeService(const std::string &aType, const std::string &aInstanceName)
 {
-    mSubscribedServices.emplace_back(*this, aType, aInstanceName);
+    auto service = MakeUnique<ServiceSubscription>(*this, aType, aInstanceName);
+
+    mSubscribedServices.push_back(std::move(service));
 
     otbrLogInfo("subscribe service %s.%s (total %zu)", aInstanceName.c_str(), aType.c_str(),
                 mSubscribedServices.size());
 
     if (aInstanceName.empty())
     {
-        mSubscribedServices.back().Browse();
+        mSubscribedServices.back()->Browse();
     }
     else
     {
-        mSubscribedServices.back().Resolve(AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, aInstanceName, aType);
+        mSubscribedServices.back()->Resolve(AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, aInstanceName, aType);
     }
 }
 
 void PublisherAvahi::UnsubscribeService(const std::string &aType, const std::string &aInstanceName)
 {
     ServiceSubscriptionList::iterator it =
         std::find_if(mSubscribedServices.begin(), mSubscribedServices.end(),
-                     [&aType, &aInstanceName](const ServiceSubscription &aService) {
-                         return aService.mType == aType && aService.mInstanceName == aInstanceName;
+                     [&aType, &aInstanceName](const std::unique_ptr<ServiceSubscription> &aService) {
+                         return aService->mType == aType && aService->mInstanceName == aInstanceName;
                      });
 
     assert(it != mSubscribedServices.end());
 
-    it->Release();
-    mSubscribedServices.erase(it);
+    {
+        std::unique_ptr<ServiceSubscription> service = std::move(*it);
+
+        mSubscribedServices.erase(it);
+        service->Release();
```

Comment:
nit: I think you can directly do following:

```
(*it)->Release();
mSubscribedServices.erase(it);
```

Reply:
The current code should be fine. Prefer to leave it as is.